### PR TITLE
Fix planning failure for mixed join and group by with grouped execution

### PR DESCRIPTION
**Pull request by haozhun**
Planning failed when
* Build side of a join contains group by (or join) that can benefit from grouped execution
* The join itself cannot take advantage of grouped execution

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 1]
```diff
@@ -470,17 +503,27 @@ private GroupedExecutionProperties processChildren(PlanNode node)
...
         private final boolean currentNodeCapable;
-        // Whether grouped execution is beneficial in the current node, or any node below it.
-        // For example, a JOIN can benefit from grouped execution because build can be flushed early, reducing peak memory requirement.
         private final boolean subTreeUseful;
 
         public GroupedExecutionProperties(boolean currentNodeCapable, boolean subTreeUseful)
         {
             this.currentNodeCapable = currentNodeCapable;
             this.subTreeUseful = subTreeUseful;
+            // Verify that "subTreeUseful -> currentNodeCapable" (->: implies)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: If you use `=>` I think it's more obvious that it means "implies" and you don't need to clarify it.

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: or use "implies" in place of the arrow:
```
// Verify that `subTreeUseful` implies `currentNodeCapable`
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 2]
```diff
@@ -419,11 +419,38 @@ protected GroupedExecutionProperties visitPlan(PlanNode node, Void context)
...
         public GroupedExecutionProperties visitJoin(JoinNode node, Void context)
         {
-            GroupedExecutionProperties properties = processChildren(node);
-            if (properties.isCurrentNodeCapable()) {
-                return new GroupedExecutionProperties(true, true);
+            GroupedExecutionProperties left = node.getLeft().accept(this, null);
+            GroupedExecutionProperties right = node.getRight().accept(this, null);
+
+            if (Optional.empty().equals(node.getDistributionType())) {
+                // This is possible when the optimizers is invoked with `forceSingleNode` set to true.
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: I think we should be able to pull `DetermineJoinDistributionType` and `DetermineSemiJoinDistributionType` outside of `if (!forceSingleNode)` check in the `PlanOptimizers`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 3]
```diff
@@ -419,11 +419,38 @@ protected GroupedExecutionProperties visitPlan(PlanNode node, Void context)
...
-            return properties;
+
+            if (Optional.of(JoinNode.DistributionType.PARTITIONED).equals(node.getDistributionType())) {
+                if (left.currentNodeCapable && right.currentNodeCapable) {
+                    return new GroupedExecutionProperties(true, true);
+                }
+                // right.subTreeUseful && !left.currentNodeCapable:
+                //   It's not particularly helpful to do grouped execution on the right side
+                //   because the benefit is likely cancelled out due to required buffering for hash build.
+                //   In theory, it could still be helpful (e.g. when the underlying aggregation's intermediate group state maybe larger than aggregation output).
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: s/maybe/may be/

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 4]
```diff
@@ -419,11 +419,38 @@ protected GroupedExecutionProperties visitPlan(PlanNode node, Void context)
...
+                //   However, this is not implemented at the moment. LookupSourceFactoryManager need to support such a lifecycle.
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: "not currently implemented"?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 5]
```diff
@@ -419,11 +419,38 @@ protected GroupedExecutionProperties visitPlan(PlanNode node, Void context)
...
+
+            if (Optional.empty().equals(node.getDistributionType())) {
+                // This is possible when the optimizers is invoked with `forceSingleNode` set to true.
+                return new GroupedExecutionProperties(false, false);
+            }
+
+            if (Optional.of(JoinNode.DistributionType.REPLICATED).equals(node.getDistributionType())) {
+                // Broadcast join maintains partitioning for the left side.
+                // Right side of a broadcast is not capable of grouped execution because it always comes from a remote exchange.
+                checkState(!right.currentNodeCapable);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: Do i understand correctly that when right side was group-exe-capable, this is cancelled by the fact there are remote exchanges on top?

Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: Yes.

This analysis is done after fragmenter. As a result, remote exchange source is a leaf node. Currently, remote exchange doesn't have the capability to have multiple streams for a single recipient task.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java (hunk 2 above)
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: Feel free to put up a PR for that.

**Pull request merged**